chisel/src/driver.rs

```diff
@@ -0,0 +1,544 @@
+                        );
+                        return;
+                    }
+                }
+            } else {
+                self.state = DriverState::Error(
+                    DriverError::MissingRequiredField(name.clone(), "file".to_string()),
+                    results,
+                );
+                return;
+            };
+
+            // Look for output path and set.
+            let output_path =
+                if let Some(output_path) = ruleset.options().get(&"output".to_string()) {
+                    chisel_debug!(1, "Found output path: {}", &output_path);
+                    PathBuf::from(output_path)
+                } else if let Some(output_path) = ruleset.options().get(&"out".to_string()) {
+                    chisel_debug!(1, "Found output path: {}", &output_path);
+                    PathBuf::from(output_path)
+                } else {
+                    chisel_debug!(1, "No output path found.");
+                    binary_path.clone()
+                };
+            rs_result.set_output_path(output_path);
+
+            // Load the wasm binary into a buffer before deserialization.
+            chisel_debug!(1, "Deserializing module from file");
+            let wasm_raw = match read(binary_path) {
+                Ok(ret) => ret,
+                Err(e) => {
+                    chisel_debug!(1, "Failed to load Wasm binary");
+                    self.state = DriverState::Error(
+                        DriverError::Internal(
+                            name.clone(),
+                            "Failed to load file".to_string(),
+                            e.into(),
+                        ),
+                        results,
+                    );
+                    return;
+                }
+            };
+
+            // Deserialize the Wasm binary and parse its names section.
+            let mut wasm = match Module::from_bytes(wasm_raw) {
+                Ok(wasm) => {
+                    chisel_debug!(1, "Successfully deserialized Wasm module");
+                    // TODO: Make this error recoverable
+                    wasm.parse_names().expect("names parsing failed")
+                }
+                Err(e) => {
+                    chisel_debug!(1, "Failed to deserialize Wasm module");
+                    self.state = DriverState::Error(
+                        DriverError::Internal(
+                            name.clone(),
+                            "Deserialization failure".to_string(),
+                            e.into(),
+                        ),
+                        results,
+                    );
+                    return;
+                }
+            };
+
+            // Consume modules in ruleset and execute.
+            while let Some((name, module)) = ruleset.modules_mut().pop_front() {
+                chisel_debug!(1, "Executing module {}", &name);
+
+                // TODO: homogeneous module handling
```

Comment:
I think the solution to this giant mess is passing a `HashMap<String, String>` as configuration as part of `ChiselModule` or `ModulePreset`.

Reply:
Attempted it in #161.